File: services/game-api/internal/core/build_operation.go
package core

// BuildOperationReplay captures the durable fields that must match before an
// existing rebuild operation may be treated as an idempotent replay.
type BuildOperationReplay struct {
	OwnerCharacterID string
	VehicleID        string
	ResultRevision   int
	ValidationHash   string
}

// ValidateBuildOperationReplay rejects operation-key reuse unless the durable
// receipt belongs to the same character/vehicle and represents exactly the
// revision/hash implied by the caller's original request.
func ValidateBuildOperationReplay(existing BuildOperationReplay, ownerCharacterID, vehicleID string, expectedRevision int, validationHash string) bool {
	if ownerCharacterID == "" || vehicleID == "" || validationHash == "" || expectedRevision < 1 {
		return false
	}
	return existing.OwnerCharacterID == ownerCharacterID &&
		existing.VehicleID == vehicleID &&
		existing.ResultRevision == expectedRevision+1 &&
		existing.ValidationHash == validationHash
}

File: src/zneondrive/build_operation.py
"""Reference-only rebuild operation replay contract."""

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class BuildOperationReplay:
    owner_character_id: str
    vehicle_id: str
    result_revision: int
    validation_hash: str


def validate_build_operation_replay(
    existing: BuildOperationReplay,
    owner_character_id: str,
    vehicle_id: str,
    expected_revision: int,
    validation_hash: str,
) -> bool:
    """Return True only for an exact payload-bound durable replay."""
    if not owner_character_id or not vehicle_id or not validation_hash or expected_revision < 1:
        return False
    return (
        existing.owner_character_id == owner_character_id
        and existing.vehicle_id == vehicle_id
        and existing.result_revision == expected_revision + 1
        and existing.validation_hash == validation_hash
    )
